neqo-transport/src/connection.rs

```diff
@@ -2677,48 +2713,48 @@ impl LossRecovery {
         (acked_packets, lost_packets)
     }
 
-    fn detect_lost_packets(&mut self, pn_space: PNSpace, cur_time: u64) -> Vec<SentPacket> {
-        self.space_mut(pn_space).loss_time = 0;
-
-        let loss_delay = (TIME_THRESHOLD
-            * (max(self.rtt_vals.latest_rtt, self.rtt_vals.smoothed_rtt) as f64))
-            as u64;
+    fn detect_lost_packets(&mut self, pn_space: PNSpace, cur_time: Instant) -> Vec<SentPacket> {
+        self.space_mut(pn_space).loss_time = None;
 
-        // Packets sent before this time are deemed lost.
-        // (cur_time < loss_delay, can happen in test)
-        let lost_send_time = if cur_time < loss_delay {
-            0
-        } else {
-            cur_time - loss_delay
-        };
-
-        // Packets with packet numbers before this are deemed lost.
-        let lost_pn = self
-            .space_mut(pn_space)
-            .largest_acked
-            .saturating_sub(PACKET_THRESHOLD);
+        let loss_delay = Duration::from_micros(
+            (TIME_THRESHOLD
+                * (max(
+                    match self.rtt_vals.smoothed_rtt {
+                        None => self.rtt_vals.latest_rtt,
+                        Some(smoothed_rtt) => max(self.rtt_vals.latest_rtt, smoothed_rtt),
+                    },
+                    GRANULARITY,
+                ))
+                .as_micros() as f64) as u64,
+        );
 
-        qdebug!(
-            [self]
-            "detect lost packets - time={}, pn={}",
-            lost_send_time,
-            lost_pn
+        let loss_deadline = cur_time - loss_delay;
+        qdebug!([self]
+            "detect lost packets = cur_time {:?} loss delay {:?} loss_deadline {:?}",
+            cur_time, loss_delay, loss_deadline
         );
 
+        // Packets with packet numbers before this are deemed lost.
         let packet_space = self.space_mut(pn_space);
 
         let mut lost = Vec::new();
         for (pn, packet) in &packet_space.sent_packets {
-            // Mark packet as lost, or set time when it should be marked.
-            if *pn <= packet_space.largest_acked {
-                if packet.time_sent <= lost_send_time || *pn <= lost_pn {
+            if Some(*pn) <= packet_space.largest_acked {
```

Comment:
This whole bit of code is a little weird.  I guess that I'm not used to seeing `Option<u64>` being used with comparison operators.

Reply:
yeah. Basically None compares as less than Some(_). It's somewhat moot because detect_lost_packets is only called after receiving an ack so largest_acked will never be None, but now at least we should feel a warm fuzzy feeling I guess for having squashed an issue where None vs Some(0) could have introduced a bug. :smiley: